### PR TITLE
BUG: Should not raise errors in .set_names for MultiIndex with nlevels == 1 (GH21149)

### LINKED ISSUE #21149: (Assuming we don't want to disallow 1-level MultiIndexes), stop identifying MultiIndex by nlevels == 1
#### Code Sample, a copy-pastable example if possible

```python
In [2]: mi = pd.MultiIndex.from_product([[0, 1]])

In [3]: mi.set_names(('first',), level=0)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-1f4ad6c7f0bf> in <module>()
----> 1 mi.set_names(('first',), level=0)

/home/nobackup/repo/pandas/pandas/core/indexes/base.py in set_names(self, names, level, inplace)
   1386 
   1387         if level is not None and self.nlevels == 1:
-> 1388             raise ValueError('Level must be None for non-MultiIndex')
   1389 
   1390         if level is not None and not is_list_like(level) and is_list_like(

ValueError: Level must be None for non-MultiIndex

```

#### Problem description

The case of ``MultiIndex`` with ``self.nlevels == 1`` is currently perfectly valid (despite e.g. ``MultiIndex.droplevel`` returning a normal ``Index`` when only one level is left). Unless we want to change this, we should always identify a ``MultiIndex`` with ``isinstance``, not by looking at ``self.nlevels``.

#### Expected Output

```python
In [4]: mi.set_names(('first',))
Out[4]: 
MultiIndex(levels=[[0, 1]],
           labels=[[0, 1]],
           names=['first'])
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: 5cf4957a2cc8429170742dd1e6435ddb8c323811
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-6-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: it_IT.UTF-8
LOCALE: it_IT.UTF-8

pandas: 0.24.0.dev0+14.g5cf4957a2.dirty
pytest: 3.5.0
pip: 9.0.1
setuptools: 39.0.1
Cython: 0.25.2
numpy: 1.14.3
scipy: 0.19.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.5.6
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2018.4
blosc: None
bottleneck: 1.2.0dev
tables: 3.3.0
numexpr: 2.6.1
feather: 0.3.1
matplotlib: 2.0.0
openpyxl: 2.3.0
xlrd: 1.0.0
xlwt: 1.3.0
xlsxwriter: 0.9.6
lxml: 4.1.1
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.0.15
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: 0.2.1

</details>

### PR DESCRIPTION
**Pull request by KalyanGokhale**
- [x] partially #21149 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Initial PR #21182 is closed - and now split in 2 different PRs

### CODE REVIEW

File: doc/source/whatsnew/v0.23.1.txt [hunk 1]
```diff
@@ -78,7 +78,8 @@ Indexing
 
 - Bug in :meth:`Series.reset_index` where appropriate error was not raised with an invalid level name (:issue:`20925`)
 - Bug in :func:`interval_range` when ``start``/``periods`` or ``end``/``periods`` are specified with float ``start`` or ``end`` (:issue:`21161`)
--
+- Bug in :meth:`MultiIndex.set_names` where error raised for a MultiIndex with self.nlevels == 1 (:issue:`21149`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.1.txt]: just say ``nlevels == 1`` (in double backticks)

File: pandas/tests/indexes/test_multi.py [hunk 2]
```diff
@@ -165,6 +165,18 @@ def test_set_name_methods(self):
         assert res is None
         assert ind.names == new_names2
 
+    def test_multiindex_set_names(self):
+        # GH 21149
+        '''Ensure that .set_names for MultiIndex with
```
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: use triple-double quotes

File: pandas/tests/indexes/test_multi.py [hunk 3]
```diff
@@ -165,6 +165,18 @@ def test_set_name_methods(self):
...
+            nlevels == 1 does not raise any errors
+        '''
+        result = pd.MultiIndex.from_product([[0, 1]])
+        result.set_names('first', level=0, inplace=True)
```
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: test both inplace=True and False with a parameterized test

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.23.1.txt (hunk 1 above)
Comment (KalyanGokhale) [reply doc/source/whatsnew/v0.23.1.txt]: Thanks - done

File: pandas/tests/indexes/test_multi.py (hunk 2 above)
Comment (KalyanGokhale) [reply pandas/tests/indexes/test_multi.py]: done

File: pandas/tests/indexes/test_multi.py (hunk 3 above)
Comment (KalyanGokhale) [reply pandas/tests/indexes/test_multi.py]: done

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @KalyanGokhale 

certainly love for you to kill the nlevels==1 checks in the rest of the code base :>

**Comment by KalyanGokhale** [comment]
Thanks @jreback - will comb through the codes for other nlevels==1 instances and would work on it

**Pull request merged**